### PR TITLE
Round 01

### PR DESCRIPTION
**Pull request by SamuelHudec**

### CODE REVIEW

File: challenge_01/base_solution.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                nn.Linear(n_in, n_out),
+                nn.BatchNorm1d(n_out),
+                nn.Dropout(p=self.dropout_p),
+                self.activation(),
+            )
+        else:
+            block = nn.Sequential(
+                nn.Linear(n_in, n_out), nn.Dropout(p=self.dropout_p), self.activation()
+            )
+        return block
```
Comment (roman-dusek) [inline challenge_01/base_solution.py]: přijde mi víc standard mít 
```
blocks = [ ]

if bn:
    blocks.append()
if dropoup:
...
blocks.append(activation)
nn.Sequantial(*blocks)
```

File: challenge_01/base_solution.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Model class follow rules by nn.Module
+    I tried to fill class by glues I got.
+    Key trick is create hidden layers by blocks follow nn.Sequential
+    """
+
+    def __init__(
+        self,
+        layer_dim_list: List[int],
+        activation: nn.modules.activation,
```
Comment (roman-dusek) [inline challenge_01/base_solution.py]: teď se spíš používá torch.relu(), atd...

nevím jestli je na to jiný typing  než Callable

File: challenge_01/base_solution.py (hunk 1 above)
Comment (roman-dusek) [reply challenge_01/base_solution.py]: přesně jak ve create_mlp

Comment (roman-dusek) [reply challenge_01/base_solution.py]: MLP je tak jednoduchá síť že se to většinou dává rovnou do create_mlp

File: challenge_01/base_solution.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        blocks = []
+        blocks.append(nn.Flatten())
+        for i in range(len(self.layer_dim_list) - 2):
+            blocks.append(
+                self._create_block(
+                    self.layer_dim_list[i], self.layer_dim_list[i + 1], self.batch_norm
+                )
+            )
+        blocks.append(nn.Linear(self.layer_dim_list[-2], self.layer_dim_list[-1]))
+        self.layer = nn.Sequential(*blocks)
```
Comment (roman-dusek) [inline challenge_01/base_solution.py]: lepší je kdyby to vracelo layer a zapíšeš to v init (to mě naučil jiřík)

File: challenge_01/data_solution.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+import torch
+import torch.nn as nn
+import torch.utils.data as data
+from torchvision import transforms
+from torchvision.datasets import CIFAR10
+
+
+class MultiLayerPerceptron(nn.Module):
+    """
```
Comment (roman-dusek) [inline challenge_01/data_solution.py]: dej si to do souboru models/mlp.py a načítej už (nebo to načítej z base solution)

File: challenge_01/data_solution.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                )
+            )
+        blocks.append(nn.Linear(self.layer_dim_list[-2], self.layer_dim_list[-1]))
+        self.layer = nn.Sequential(*blocks)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.layer(x)
+
+
+def training_loop(model: nn.Module, dataloader: data.DataLoader, lr: float):
```
Comment (roman-dusek) [inline challenge_01/data_solution.py:10]: do souboru train_utils.py

přidej možnost argument pro optimizer a loss (nebo je rovnou inicializuj v modelu)

File: challenge_01/data_solution.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    torch.manual_seed(538)
+
+    # Prepare CIFAR-10 dataset
+    dataset = CIFAR10(os.getcwd(), download=True, transform=transforms.ToTensor())
+
+    # feed training loop
+    n_featers = 32 * 32 * 3
+    model = MultiLayerPerceptron([n_featers, 80, 40, 10], nn.ReLU, True, 0.1)
+    dataloader = data.DataLoader(dataset, batch_size=100, shuffle=True, num_workers=1)
+    training_loop(model, dataloader, lr=0.1)
```
Comment (roman-dusek) [inline challenge_01/data_solution.py]: zkoušel jsi udělat větší počet neuronů ve vrstvách aby šla loss dolu?

nebo zkus Adam optimizer

Comment (roman-dusek) [reply challenge_01/data_solution.py]: CIFAR je jinak docela těžký na MLP si myslím

to spíš zkus MNIST

Comment (roman-dusek) [reply challenge_01/data_solution.py]: dobrá praxe je vzít si jednu batch nebo 10 samplů z batche a projet na tom 100 epoch, jestli se ti to model na trénuje, pokud ano implementace je správná (to se dá testovat i na random datech)

File: challenge_01/base_solution.py (hunk 1 above)
Comment (SamuelHudec) [reply challenge_01/base_solution.py]: Aj mne sa zdal ten if else kostrbatý dik za radu je to tam ;) hej je jednoduchá ale dal si na vstup list a ten list môže byť rozne velký s roznimi parametrami, tak som to urobil modulárne

File: challenge_01/base_solution.py (hunk 2 above)
Comment (SamuelHudec) [reply challenge_01/base_solution.py]: s týmto som si nevedel rady tak som tam hodil intuitivne čo prislo vhod podla triedy od kial pochádza aktivačná funkcia ale je pravda že tam môžeš dať potom aj niejaku vlastnu a už to bude držkovať, tak som to zmenil

File: challenge_01/base_solution.py (hunk 3 above)
Comment (SamuelHudec) [reply challenge_01/base_solution.py]: ked to hovorí jiřík tak je to pravda!

File: challenge_01/data_solution.py (hunk 4 above)
Comment (SamuelHudec) [reply challenge_01/data_solution.py]: to ma presne napadlo teraz :D

File: challenge_01/data_solution.py (hunk 5 above)
Comment (SamuelHudec) [reply challenge_01/data_solution.py:10]: jedná sa o výzvu nie o knižnicu budem rád kod písať dookola nech ho viem na spamat ;)

File: challenge_01/data_solution.py (hunk 6 above)
Comment (SamuelHudec) [reply challenge_01/data_solution.py]: super diky!
